### PR TITLE
M106 Fan Speed New Features - rePATCHED ;-)

### PR DESCRIPTION
**Pull request by ghost**
:white_check_mark:**FAN SPEED 'New Features'**

> Now you can use your own fan speed during a print and recover the original speed
> Usefull when needed to go on parking with a special fan speed value or anything need a special blowing property

LCD settings in  Filament' menu
M106 - Fan on: S0-255 : T3-255 / T2 / T1 Temporary set speed during printing
T1 return to previous / T2 apply extra speed / T3-255 set without applying

### CODE REVIEW

**Comment by thinkyhead** [comment]
It's a fine concept, but as it costs SRAM it needs to be made an optional feature.

**Review by ghost — COMMENTED** [review]
New adds
Enable/disable feature
LCD Settings Of Extra Fanspeed 
En definition
Fr definition
And cleaning

File: Marlin/Marlin_main.cpp [hunk 1]
```diff
@@ -7435,14 +7439,31 @@ inline void gcode_M105() {
...
     uint16_t s = parser.ushortval('S', 255);
-    NOMORE(s, 255);
     const uint8_t p = parser.byteval('P', 0);
-    if (p < FAN_COUNT) fanSpeeds[p] = s;
-  }
+    if (p < FAN_COUNT) {
+    #if ENABLED(EXTRA_FANSPEED)      
+      uint16_t t = parser.ushortval('T', 0);
+      if(t>0) {
+              if (t>2){NOMORE(t, 255);new_fanSpeeds[p]=t; }
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: This should be `NOMORE(t, FAN_COUNT - 1)`

File: Marlin/Marlin.h [hunk 2]
```diff
@@ -355,6 +355,10 @@ extern float soft_endstop_min[XYZ], soft_endstop_max[XYZ];
 
 #if FAN_COUNT > 0
   extern int16_t fanSpeeds[FAN_COUNT];
+  #if ENABLED(EXTRA_FANSPEED)      
+    extern int16_t old_fanSpeeds[FAN_COUNT];
+    extern int16_t new_fanSpeeds[FAN_COUNT];
```
Comment (thinkyhead) [inline Marlin/Marlin.h]: …or this should be `[256]`…

**Comment by thinkyhead** [comment]
Technically this is fine. But I'm struggling to find a use-case where the host or G-code generator wouldn't know what fan speeds it wants to set / restore. Marlin is —as much as possible— a deterministic machine. Why would it need to "memorize" a fan speed in any context, when clearly whatever speed it would be restoring is already well-known by the machine controller?

File: Marlin/Marlin.h [hunk 3]
```diff
@@ -355,6 +355,10 @@ extern float soft_endstop_min[XYZ], soft_endstop_max[XYZ];
 
 #if FAN_COUNT > 0
   extern int16_t fanSpeeds[FAN_COUNT];
+  #if ENABLED(EXTRA_FANSPEED)
```
Comment (thinkyhead) [inline Marlin/Marlin.h:358]: `EXTRA_FANSPEED` or `EXTRA_FAN_SPEED`?

Comment (ghost) [reply Marlin/Marlin.h:358]: Fixed ;-)

**Review by ghost — COMMENTED** [review]
It's merged , you can close
Thks 👍 
Merge my Firmware retract ' you can trust , i use it every day' fully debbuged
Your actual fwretract have some issue ' look in the issue list ' some have the zlift that not recover , my Fwretract solve all
Priority to the ' good sense ' , G10S1 protected , zlift too , and more
Thks

**Pull request closed without merging**